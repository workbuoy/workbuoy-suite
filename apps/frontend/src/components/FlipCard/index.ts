export { default, default as FlipCard } from "./FlipCard";
export type { FlipCardProps, FlipCardSize, FlipCardStrings } from "./FlipCard";
