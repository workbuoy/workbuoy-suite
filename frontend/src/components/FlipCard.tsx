import React, { useEffect, useMemo, useState } from "react";
import BuoyChat from "../features/buoy/BuoyChat";
import NaviGrid from "../features/navi/NaviGrid";
import { IntrospectionBadge } from "./IntrospectionBadge";
type Mode = "CHAT" | "NAVI";
export default function FlipCard() {
  const [mode, setMode] = useState<Mode>("CHAT");
  const flipped = mode === "NAVI";
  useEffect(() => {
    const onKey = (e: KeyboardEvent) => {
      if (e.key === " " || e.key === "Enter") { e.preventDefault(); setMode(m => m==="CHAT" ? "NAVI" : "CHAT"); }
    };
    window.addEventListener("keydown", onKey);
    return () => window.removeEventListener("keydown", onKey);
  }, []);
  const ariaFront = useMemo(()=>({ "aria-hidden": flipped, "aria-label": "Buoy chat" }), [flipped]);
  const ariaBack  = useMemo(()=>({ "aria-hidden": !flipped, "aria-label": "Navi oversikt" }), [flipped]);
  return (
    <div className="perspective" style={{position:"relative", height:"min(82vh, 760px)"}}>
      <div className="flipcard cardbg" style={{position:"absolute", inset:0, borderRadius:16, transform: `rotateY(${flipped?180:0}deg)`}}>
        <section {...ariaFront} className="flipface" style={{padding:12}}>
          <Header side="Buoy" onFlip={()=>setMode("NAVI")} />
          <div style={{position:"absolute", inset:"56px 12px 12px 12px"}}><BuoyChat/></div>
        </section>
        <section {...ariaBack} className="flipface" style={{transform:"rotateY(180deg)", padding:12}}>
          <Header side="Navi" onFlip={()=>setMode("CHAT")} />
          <div style={{position:"absolute", inset:"56px 12px 12px 12px"}}><NaviGrid/></div>
        </section>
      </div>
    </div>
  );
}
function Header({side, onFlip}:{side:"Buoy"|"Navi"; onFlip:()=>void}) {
  return (
    <div style={{display:"flex", alignItems:"center", gap:10}}>
      <strong style={{letterSpacing:.3, opacity:.9}}>{side}</strong>
      <span style={{flex:1}}/>
      <IntrospectionBadge/>
      <HealthBadge/>
      <button onClick={onFlip} aria-label={side==="Buoy"?"Gå til Navi":"Gå til Buoy"} className="chip" style={{background:"transparent"}}>Flip</button>
    </div>
  );
}
function HealthBadge() {
  const [state, setState] = useState<"ok"|"wait"|"err">("wait");
  useEffect(() => {
    fetch("/api/health").then(r=>r.ok?r.json():Promise.reject()).then(()=>setState("ok")).catch(()=>setState("err"));
  }, []);
  const color = state==="ok"?"var(--ok)": state==="wait"?"var(--warn)":"var(--err)";
  const label = state==="ok"?"Backend OK": state==="wait"?"Sjekker…":"Feil";
  return <span className="chip" style={{borderColor:color, color}}>{label}</span>;
}