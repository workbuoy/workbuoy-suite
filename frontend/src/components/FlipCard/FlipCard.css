.flip-card-host {
  position: relative;
  margin: 0 auto;
  will-change: transform;
  transform: translateZ(0);
}

.flip-card {
  position: relative;
  height: 100%;
  width: 100%;
  border-radius: var(--radius-lg);
  transform-style: preserve-3d;
  transition: transform var(--transition-slow);
  overflow: hidden;
  will-change: transform;
}

.flip-card[data-motion="calm"] {
  transition: transform 420ms cubic-bezier(0.22, 0.8, 0.26, 1);
}

.flip-card--flipped {
  transform: rotateY(180deg);
}

.flip-card[data-style="fast"] {
  transform-style: flat;
  transition: none;
}

.flip-card[data-style="fast"].flip-card--flipped {
  transform: none;
}

.flip-card[data-style="fast"] .flip-card-face {
  backface-visibility: visible;
  opacity: 0;
  transform: translateY(8px);
  transition: opacity 220ms ease, transform 220ms ease;
  will-change: opacity, transform;
}

.flip-card[data-style="fast"][data-motion="calm"] .flip-card-face {
  transition-duration: 280ms;
  transition-timing-function: cubic-bezier(0.24, 0.82, 0.25, 1);
}

.flip-card[data-style="fast"][data-side="front"] .flip-card-face--front,
.flip-card[data-style="fast"][data-side="back"] .flip-card-face--back {
  opacity: 1;
  transform: translateY(0);
}

.flip-card[data-style="fast"][data-side="front"] .flip-card-face--back {
  opacity: 0;
  transform: translateY(12px);
}

.flip-card[data-style="fast"][data-side="back"] .flip-card-face--front {
  opacity: 0;
  transform: translateY(-12px);
}

.flip-card-toolbar {
  position: absolute;
  inset: 0 0 auto 0;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: var(--space-md) var(--space-lg);
  z-index: var(--z-overlay);
  pointer-events: none;
}

.flip-card-toolbar__side,
.flip-card-toolbar__actions {
  display: flex;
  align-items: center;
  gap: 8px;
  pointer-events: auto;
  flex-wrap: wrap;
}

.flip-card-toolbar__flip,
.flip-card-toolbar__connect,
.flip-card-toolbar__resize {
  background: rgba(12, 16, 24, 0.65);
  border-color: var(--stroke-subtle);
  color: var(--fg-default);
}

.flip-card-toolbar__resize {
  cursor: ew-resize;
}

.flip-card-face {
  position: absolute;
  inset: 0;
  backface-visibility: hidden;
  padding: calc(var(--space-xl) * 2) var(--space-md) var(--space-md);
  display: flex;
}

.flip-card-face--front {
  transform: rotateY(0deg);
}

.flip-card-face--back {
  transform: rotateY(180deg);
}

.flip-face-content {
  flex: 1;
  display: flex;
  flex-direction: column;
  height: 100%;
}

.flip-card-connect {
  position: absolute;
  inset: 0;
  background: rgba(5, 8, 12, 0.72);
  display: grid;
  place-items: center;
  z-index: var(--z-dialog);
}

.flip-card-connect__form {
  background: var(--surface-elevated);
  border: 1px solid var(--stroke-hairline);
  border-radius: var(--radius-md);
  padding: var(--space-xl);
  width: min(320px, 90vw);
  display: grid;
  gap: var(--space-lg);
}

.flip-card-connect__label {
  display: grid;
  gap: 6px;
  font-size: 0.85rem;
}

.flip-card-connect__label input,
.flip-card-connect__label select {
  background: rgba(12, 16, 24, 0.65);
  border: 1px solid var(--stroke-subtle);
  border-radius: var(--radius-md);
  padding: var(--space-sm) var(--space-md);
  color: var(--fg-default);
}

.flip-card-connect__actions {
  display: flex;
  gap: var(--space-md);
  justify-content: flex-end;
}

@media (prefers-reduced-motion: reduce) {
  .flip-card {
    transition: none;
  }

  .flip-card[data-style="fast"] .flip-card-face {
    transition: none;
  }
}

@media (max-width: 720px) {
  .flip-card-face {
    padding: calc(var(--space-xl) * 1.75) var(--space-sm) var(--space-sm);
  }

  .flip-card-toolbar {
    padding: var(--space-sm) var(--space-md);
  }
}

@media (max-width: 540px) {
  .flip-card-toolbar {
    flex-direction: column;
    align-items: stretch;
    gap: 6px;
  }

  .flip-card-toolbar__actions {
    justify-content: flex-end;
  }
}
