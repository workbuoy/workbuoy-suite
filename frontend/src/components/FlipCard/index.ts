export { default, default as FlipCard } from "./FlipCard";
export type { FlipCardProps, FlipCardSize } from "./FlipCard";
