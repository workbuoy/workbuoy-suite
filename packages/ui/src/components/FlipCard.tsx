import type { CSSProperties, KeyboardEvent, ReactNode } from "react";
import { useCallback, useEffect, useMemo, useState } from "react";
import type { MotionProps } from "framer-motion";
import { motion } from "framer-motion";

export type FlipCardProps = {
  front: ReactNode;
  back: ReactNode;
  isFlipped?: boolean;
  onFlip?: () => void;
  className?: string;
};

const baseCardClass = "wbui-flip-card wbui-focus-ring";
const faceClass = "wbui-flip-card__face";

function mergeClassName(extra?: string) {
  return extra ? `${baseCardClass} ${extra}` : baseCardClass;
}

const baseCardStyle: CSSProperties = {
  position: "relative",
  display: "grid",
  cursor: "pointer",
  transformStyle: "preserve-3d",
};

const frontStyle: CSSProperties = {
  gridArea: "1 / 1",
  backfaceVisibility: "hidden",
};

const backStyle: CSSProperties = {
  ...frontStyle,
  transform: "rotateY(180deg)",
};

const wrapperStyle: CSSProperties = {
  perspective: "1200px",
  display: "inline-block",
};

export default function FlipCard({
  front,
  back,
  isFlipped,
  onFlip,
  className,
}: FlipCardProps) {
  const [internalFlipped, setInternalFlipped] = useState(false);
  const [prefersReducedMotion, setPrefersReducedMotion] = useState(false);

  const resolvedFlipped = isFlipped ?? internalFlipped;

  useEffect(() => {
    if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
      return;
    }

    const mediaQueryList = window.matchMedia("(prefers-reduced-motion: reduce)");
    setPrefersReducedMotion(mediaQueryList.matches);

    const listener = (event: MediaQueryListEvent) => {
      setPrefersReducedMotion(event.matches);
    };

    if (typeof mediaQueryList.addEventListener === "function") {
      mediaQueryList.addEventListener("change", listener);
    } else if (typeof mediaQueryList.addListener === "function") {
      mediaQueryList.addListener(listener);
    }

    return () => {
      if (typeof mediaQueryList.removeEventListener === "function") {
        mediaQueryList.removeEventListener("change", listener);
      } else if (typeof mediaQueryList.removeListener === "function") {
        mediaQueryList.removeListener(listener);
      }
    };
  }, []);

  const motionDuration = prefersReducedMotion ? 0 : 0.25;

  const transition: MotionProps["transition"] = useMemo(
    () => ({
      duration: motionDuration,
      ease: [0.45, 0, 0.15, 1],
    }),
    [motionDuration],
  );

  const toggle = useCallback(() => {
    if (isFlipped === undefined) {
      setInternalFlipped((value) => !value);
    }

    onFlip?.();
  }, [isFlipped, onFlip]);

  const handleKeyDown = useCallback(
    (event: KeyboardEvent<HTMLDivElement>) => {
      if (event.key === "Enter" || event.key === " " || event.key === "Space" || event.key === "Spacebar") {
        event.preventDefault();
        toggle();
      }
    },
    [toggle],
  );

  const cardClassName = useMemo(() => mergeClassName(className), [className]);

  return (
    <div style={wrapperStyle}>
      <motion.div
        role="button"
        tabIndex={0}
        aria-pressed={resolvedFlipped}
        data-flipped={resolvedFlipped}
        className={cardClassName}
        style={{
          ...baseCardStyle,
          transition: prefersReducedMotion
            ? "none"
            : `transform ${motionDuration}s ease`,
        }}
        animate={{ rotateY: resolvedFlipped ? 180 : 0 }}
        initial={false}
        transition={transition}
        onClick={toggle}
        onKeyDown={handleKeyDown}
      >
        <div
          className={`${faceClass} ${faceClass}--front`}
          style={frontStyle}
          aria-hidden={resolvedFlipped}
        >
          {front}
        </div>
        <div
          className={`${faceClass} ${faceClass}--back`}
          style={backStyle}
          aria-hidden={!resolvedFlipped}
        >
          {back}
        </div>
      </motion.div>
    </div>
  );
}
