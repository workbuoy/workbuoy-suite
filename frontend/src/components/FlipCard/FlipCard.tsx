import React, { useCallback, useMemo, useState, useEffect } from "react";
import BuoyChat from "@/features/buoy/BuoyChat";
import NaviGrid from "@/features/navi/NaviGrid";
import { IntrospectionBadge } from "@/components/IntrospectionBadge";

type Mode = "CHAT" | "NAVI";

type FlipCardProps = {
  onConnect?: () => void;
};

export default function FlipCard({ onConnect }: FlipCardProps = {}) {
  const [mode, setMode] = useState<Mode>("CHAT");
  const [expanded, setExpanded] = useState(false);

  const flipped = mode === "NAVI";

  const ariaFront = useMemo(
    () => ({ "aria-hidden": flipped, "aria-label": "Buoy chat" }),
    [flipped],
  );
  const ariaBack = useMemo(
    () => ({ "aria-hidden": !flipped, "aria-label": "Navi oversikt" }),
    [flipped],
  );

  const handleFlip = useCallback(() => {
    setMode((current) => (current === "CHAT" ? "NAVI" : "CHAT"));
  }, []);

  const handleResize = useCallback(() => {
    setExpanded((current) => !current);
  }, []);

  const handleKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLDivElement>) => {
      if (event.target !== event.currentTarget) {
        return;
      }

      if (event.key === " " || event.key === "Enter") {
        event.preventDefault();
        handleFlip();
        return;
      }

      if (event.key.toLowerCase() === "r") {
        event.preventDefault();
        handleResize();
      }
    },
    [handleFlip, handleResize],
  );

  return (
    <div
      className="perspective"
      role="application"
      aria-label="Workbuoy flipcard"
      tabIndex={0}
      onKeyDown={handleKeyDown}
      style={{
        position: "relative",
        height: expanded ? "min(92vh, 840px)" : "min(82vh, 760px)",
      }}
    >
      <div
        className="flipcard cardbg"
        style={{
          position: "absolute",
          inset: 0,
          borderRadius: 16,
          transform: `rotateY(${flipped ? 180 : 0}deg)`,
        }}
      >
        <section {...ariaFront} className="flipface" style={{ padding: 12 }}>
          <Header
            side="Buoy"
            onFlip={handleFlip}
            onResize={handleResize}
            isExpanded={expanded}
            onConnect={onConnect}
          />
          <div style={{ position: "absolute", inset: "56px 12px 12px 12px" }}>
            <BuoyChat />
          </div>
        </section>
        <section
          {...ariaBack}
          className="flipface"
          style={{ transform: "rotateY(180deg)", padding: 12 }}
        >
          <Header
            side="Navi"
            onFlip={handleFlip}
            onResize={handleResize}
            isExpanded={expanded}
          />
          <div style={{ position: "absolute", inset: "56px 12px 12px 12px" }}>
            <NaviGrid />
          </div>
        </section>
      </div>
    </div>
  );
}

type HeaderProps = {
  side: "Buoy" | "Navi";
  onFlip: () => void;
  onResize: () => void;
  isExpanded: boolean;
  onConnect?: () => void;
};

function Header({ side, onFlip, onResize, isExpanded, onConnect }: HeaderProps) {
  const flipLabel = side === "Buoy" ? "Gå til Navi" : "Gå til Buoy";
  const resizeLabel = isExpanded ? "Reduser kort" : "Utvid kort";

  return (
    <div style={{ display: "flex", alignItems: "center", gap: 10 }}>
      <strong style={{ letterSpacing: 0.3, opacity: 0.9 }}>{side}</strong>
      <span style={{ flex: 1 }} />
      <IntrospectionBadge />
      <HealthBadge />
      {onConnect ? <ConnectButton onConnect={onConnect} /> : null}
      <button
        type="button"
        onClick={onResize}
        aria-label={resizeLabel}
        className="chip"
      >
        Resize
      </button>
      <button
        type="button"
        onClick={onFlip}
        aria-label={flipLabel}
        className="chip"
        style={{ background: "transparent" }}
      >
        Flip
      </button>
    </div>
  );
}

type ConnectButtonProps = {
  onConnect: () => void;
};

function ConnectButton({ onConnect }: ConnectButtonProps) {
  const handleClick = useCallback(() => {
    onConnect();
  }, [onConnect]);

  const handleKeyDown = useCallback(
    (event: React.KeyboardEvent<HTMLButtonElement>) => {
      if (event.key === " " || event.key === "Enter") {
        event.preventDefault();
        onConnect();
      }
    },
    [onConnect],
  );

  return (
    <button
      type="button"
      className="chip"
      onClick={handleClick}
      onKeyDown={handleKeyDown}
    >
      Connect
    </button>
  );
}

type HealthState = "ok" | "wait" | "err";

function HealthBadge() {
  const [state, setState] = useState<HealthState>("wait");

  useEffect(() => {
    let cancelled = false;
    fetch("/api/health")
      .then((response) => (response.ok ? response.json() : Promise.reject()))
      .then(() => {
        if (!cancelled) setState("ok");
      })
      .catch(() => {
        if (!cancelled) setState("err");
      });
    return () => {
      cancelled = true;
    };
  }, []);

  const color =
    state === "ok"
      ? "var(--ok)"
      : state === "wait"
        ? "var(--warn)"
        : "var(--err)";
  const label =
    state === "ok" ? "Backend OK" : state === "wait" ? "Sjekker…" : "Feil";

  return (
    <span className="chip" style={{ borderColor: color, color }}>
      {label}
    </span>
  );
}
