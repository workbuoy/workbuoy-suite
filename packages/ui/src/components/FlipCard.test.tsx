import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";

import FlipCard from "./FlipCard";

describe("FlipCard", () => {
  it("toggles aria-pressed on click when uncontrolled", async () => {
    const user = userEvent.setup();
    render(<FlipCard front={<div>Front</div>} back={<div>Back</div>} />);

    const card = screen.getByRole("button", { name: /front/i });

    expect(card).toHaveAttribute("aria-pressed", "false");

    await act(async () => {
      await user.click(card);
    });

    expect(card).toHaveAttribute("aria-pressed", "true");

    await act(async () => {
      await user.click(card);
    });

    expect(card).toHaveAttribute("aria-pressed", "false");
  });

  it("calls onFlip for keyboard interactions", async () => {
    const onFlip = vi.fn();
    const user = userEvent.setup();

    render(<FlipCard front={<div>Front</div>} back={<div>Back</div>} onFlip={onFlip} />);

    await act(async () => {
      await user.tab();
    });

    const card = screen.getByRole("button", { name: /front/i });
    expect(card).toHaveFocus();

    await act(async () => {
      await user.keyboard("{Enter}");
    });
    await act(async () => {
      await user.keyboard(" ");
    });

    expect(onFlip).toHaveBeenCalledTimes(2);
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <FlipCard
        front={<div>Front Snapshot</div>}
        back={<div>Back Snapshot</div>}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
