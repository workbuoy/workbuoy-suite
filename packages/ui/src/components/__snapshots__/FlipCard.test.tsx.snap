// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FlipCard > matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="perspective: 1200px; display: inline-block;"
  >
    <div
      aria-pressed="false"
      class="wbui-flip-card"
      data-flipped="false"
      role="button"
      style="position: relative; display: grid; cursor: pointer; transform-style: preserve-3d; transform: none;"
      tabindex="0"
    >
      <div
        aria-hidden="false"
        class="wbui-flip-card__face wbui-flip-card__face--front"
        style="grid-area: 1 / 1; backface-visibility: hidden;"
      >
        <div>
          Front Snapshot
        </div>
      </div>
      <div
        aria-hidden="true"
        class="wbui-flip-card__face wbui-flip-card__face--back"
        style="grid-area: 1 / 1; backface-visibility: hidden; transform: rotateY(180deg);"
      >
        <div>
          Back Snapshot
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
